src/evict/evict_lru.c

```diff
@@ -1662,8 +1662,15 @@ __evict_walk_file(WT_SESSION_IMPL *session,
 		FLD_SET(walk_flags, WT_READ_PREV);
 
 	/*
-	 * Get some more eviction candidate pages.
-	 *
+	 * Get some more eviction candidate pages, starting at the last saved
+	 * point. Clear the saved point immediately, we assert when discarding
+	 * pages we're not discarding an eviction point, so this clear must be
+	 * complete before the page is released.
+	 */
+	ref = btree->evict_ref;
+	WT_PUBLISH(btree->evict_ref, NULL);
```

Comment:
I retract my comment about line 783, the fact this thread is the one to discard the page means it's OK.

I still think the barrier is required at line 1671, I don't see any guarantee that setting `btree.evict_ref` won't be delayed until after another thread tests in `__wt_ref_out` that `btree.evict_ref != the ref being discarded`.